### PR TITLE
[Fix] C++ API SetOutputShape for register custom op.

### PR DESCRIPTION
**Pull request by mingyueliuh**
### Description
Bug fix  for the SetOutputShape method in custom op shape inference.

### Motivation and Context

- Bug a :   A obvious bug that will cause all dimensions to be 1. 
https://github.com/microsoft/onnxruntime/blob/main/include/onnxruntime/core/session/onnxruntime_cxx_inline.h#L2014
integer_dims.push_back(dim.IsInt()); ->  integer_dims.push_back(dim.AsInt());

- Bug b :  vector out of range error
op's input maybe a scalar and shape is empty.
https://github.com/microsoft/onnxruntime/blob/main/include/onnxruntime/core/session/onnxruntime_cxx_inline.h#L1985

### CODE REVIEW

**Comment by mingyueliuh** [comment]
@jywu-msft please take a look when you are available. Thanks.

File: include/onnxruntime/core/session/onnxruntime_cxx_inline.h [hunk 1]
```diff
@@ -1996,9 +1998,10 @@ inline ShapeInferContext::ShapeInferContext(const OrtApi* ort_api,
   }
 }
 
-inline Status ShapeInferContext::SetOutputShape(size_t indice, const Shape& shape) {
+inline Status ShapeInferContext::SetOutputShape(size_t indice, const Shape& shape, ONNXTensorElementDataType type) {
   OrtTensorTypeAndShapeInfo* info = {};
   ORT_CXX_RETURN_ON_API_FAIL(ort_api_->CreateTensorTypeAndShapeInfo(&info));
+  ORT_CXX_RETURN_ON_API_FAIL(ort_api_->SetTensorElementType(info, type));
```
Comment (adrianlizarraga) [inline include/onnxruntime/core/session/onnxruntime_cxx_inline.h:2004]: I believe the underlying C API (ShapeInferContext_SetOutputTypeShape) currently ignores the tensor element type, so this change alone will not have any effect. 

We would probably need to update the C API to use the element type, and we would need to add unit tests to verify the functionality. Please refer to:
https://github.com/microsoft/onnxruntime/blob/218301403d76c3dbd7db4731071876273dfaa40d/onnxruntime/core/session/custom_ops.cc#L346-L356

https://github.com/microsoft/onnxruntime/blob/218301403d76c3dbd7db4731071876273dfaa40d/onnxruntime/core/session/custom_ops.cc#L93-L109

**Review by adrianlizarraga — COMMENTED** [review]
Thanks for contributing! Please refer to my comment regarding the new element type argument. Due to the potential complexity of implementing the change correctly, I would recommend separating that out into a separate pull request. Perhaps this pull request could only include the two bug fixes instead.

What do you think?

File: include/onnxruntime/core/session/onnxruntime_cxx_inline.h (hunk 1 above)
Comment (mingyueliuh) [reply include/onnxruntime/core/session/onnxruntime_cxx_inline.h:2004]: Make sense ！

**Comment by mingyueliuh** [comment]
@adrianlizarraga  This PR has been updated and only includes two bug fixes. 
And I will raise a new PR for support specify tensor element type after this PR is merged.

**Comment by mingyueliuh** [comment]
@adrianlizarraga  Is there any other concern with this PR?

@jywu-msft  The PRs related aligns with your suggestion (https://github.com/microsoft/onnxruntime/pull/21280#issuecomment-2218124149) in the other PR submitted to VitisAI EP, where the custom op implementation should be included within the EP rather than in the public API custom ops can completely within vitis ai ep implementation.

**Comment by adrianlizarraga** [comment]
@mingyueliuh I'm running CI to make sure all tests pass.

**Review by adrianlizarraga — APPROVED** [review]

**Pull request merged**